### DISCUSSION TITLE
Feature request: Correct cluster support

### DISCUSSION THREAD

**Discussion by Munsio** [dougiteixeira/proxmoxve#118] [Feature request] [answered]
First of all! Thank you for that great component, the built in proxmox support of HA is not really that good and I am looking forward to updates and more for this one.

For the feature: I am running a 5 node cluster which hosts my whole homelab (including homeassistant). When I add the URL the "node" selector is only able to choose 1 node instead of selecting multiple.

It does work when I add each not separately but that also increases the traffic 5 times. So my suggestion would be to allow multiple nodes with one integration.

If you want it then I can do the implementation and send you a PR for it!?

**Comment by dougiteixeira** [comment]
At the moment I don't have the time to invest in these improvements, but I'm happy to merge them if you want to contribute.

**Comment by dougiteixeira** [comment]
Ei @Munsio, I made some changes that allow selecting multiple nodes and the VM/container to be automatically updated on the running node.

But I don't have clusters configured, can you test the beta version [2.0.0.b0](https://github.com/dougiteixeira/proxmoxve/releases/tag/2.0.0.b0) and give feedback on how it works?

**Comment by Munsio** [comment]
Sure will do as soon as I get back home and will update you if it is working or not.

**Comment by Munsio** [comment]
I removed all the integrations except one and tried to reconfigure it but as soon as I choose "Add or remove Nodes...." I get an Error box without text and this in the logs:
```
This error originated from a custom integration.

Logger: aiohttp.server
Source: custom_components/proxmoxve/config_flow.py:317
Integration: Proxmox VE (documentation, issues)
First occurred: 18:37:51 (1 occurrences)
Last logged: 18:37:51

Error handling request
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/aiohttp/web_protocol.py", line 433, in _handle_request
    resp = await request_handler(request)
  File "/usr/local/lib/python3.10/site-packages/aiohttp/web_app.py", line 504, in _handle
    resp = await handler(request)
  File "/usr/local/lib/python3.10/site-packages/aiohttp/web_middlewares.py", line 117, in impl
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/security_filter.py", line 85, in security_filter_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/forwarded.py", line 227, in forwarded_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/request_context.py", line 28, in request_context_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/ban.py", line 80, in ban_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/auth.py", line 235, in auth_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/view.py", line 146, in handle
    result = await result
  File "/usr/src/homeassistant/homeassistant/components/config/config_entries.py", line 241, in post
    return await super().post(request, flow_id)
  File "/usr/src/homeassistant/homeassistant/components/http/data_validator.py", line 72, in wrapper
    result = await method(view, request, data, *args, **kwargs)
  File "/usr/src/homeassistant/homeassistant/helpers/data_entry_flow.py", line 110, in post
    result = await self._flow_mgr.async_configure(flow_id, data)
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 267, in async_configure
    result = await self._async_handle_step(
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 367, in _async_handle_step
    result: FlowResult = await getattr(flow, method)(user_input)
  File "/config/custom_components/proxmoxve/config_flow.py", line 317, in async_step_change_expose
    ] = f"{resource['vmid']} {resource['name']}"
KeyError: 'name'
```

**Comment by dougiteixeira** [comment]
Can you enable debug logs and send here?

Does any VM of yours not have the name configured?

**Comment by Munsio** [comment]
I deleted the whole integration and tried to add it again but I do have the same problem.

Afaik debug log can only be enabled when the integration is added but not during the setup wizard.

One of my servers in the cluster is powered down as it uses too much watts when it is running all the time. There the vms may have no name.

**Comment by dougiteixeira** [comment]
Include this in your config file:

```
logger:
  default: warning
  logs:
    custom_components.proxmoxve: debug
    proxmoxer: debug
```

**Comment by Munsio** [comment]
Even with a restart i got the same stacktrace as above.

**Comment by dougiteixeira** [comment]
I made some changes that I think might solve this error (at least in the configuration part).

You can test version [2.0.0.b1](https://github.com/dougiteixeira/proxmoxve/releases/tag/2.0.0.b1) and give feedback.

**Comment by Munsio** [comment]
Now it is working, I was able to add the integration:

The node list could use a dedicated headline:
![image](https://github.com/dougiteixeira/proxmoxve/assets/646513/2610205d-75a3-4103-90eb-0e8c0be05fa4)

But I am a little bit unsure what this section should be:
![image](https://github.com/dougiteixeira/proxmoxve/assets/646513/623afa40-995a-4c6a-8909-340d9178bcdd)

It seems that where usual the name of the VM/LXC now it only shows "Unnamed Service"
![image](https://github.com/dougiteixeira/proxmoxve/assets/646513/d8261d53-2ef6-4da4-a47e-a2776018451e)

I do have 11 VMs - 9 active / 2 offline. But I only have 7 services.
Should LXC Containers also be exposed as a service?
Should Nodes be exposed as a service?

The rest looks good.

**Comment by Munsio** [comment]
One feature request maybe.

It would be great if I could set multiple hostnames to avoid a loss of connection when one of the nodes is not available. The credentials are the same for every host as I am using the "pve" realm for authentication on the API.

**Comment by dougiteixeira** [comment]
> It seems that where usual the name of the VM/LXC now it only shows "Unnamed Service"

This happened to me, refreshing the page did the trick... if it still doesn't have a name let me know.



> I do have 11 VMs - 9 active / 2 offline. But I only have 7 services.
> Should LXC Containers also be exposed as a service?
> Should Nodes be exposed as a service?

All that you select in the configuration should appear as service.

If it no longer exists on the server, a repair will be created informing you to change the settings.

**Comment by dougiteixeira** [comment]
> One feature request maybe.
> 
> It would be great if I could set multiple hostnames to avoid a loss of connection when one of the nodes is not available. The credentials are the same for every host as I am using the "pve" realm for authentication on the API.

This is a little difficult to deal with, I'll see what's feasible.

**Comment by dougiteixeira** [comment]
> I do have 11 VMs - 9 active / 2 offline. But I only have 7 services.
> Should LXC Containers also be exposed as a service?
> Should Nodes be exposed as a service?

See if refreshing the page shows all the services.

**Comment by dougiteixeira** [comment]
> The node list could use a dedicated headline:

This was a migration error from the previous form to the current one... it was fixed in #85.

**Comment by dougiteixeira** [comment]
If possible, I would like you to perform a test.

The integration checks the node that the VM/container is running through the API query in `/cluster/resources`, but I don't know the behavior when a VM/container is running only one node and this node goes offline.

I want to figure out the behavior to handle this and not break the whole integration.

Can you perform such tests please?

**Comment by Munsio** [comment]
About the missing services:
I did use the wrong user with the wrong credentials, once I changed that and the extension did re-initialize itself everything showed up except the node/VM/LXC that are unavailable during that time (due to the node being offline)

For the test:
The node I am authenticated to is: "pve-asus1"
I turned off the node "pve-main" with the extension (shutdown)
All sensors of the VMs + Node of "pve-main" switched to "unavailable"
When I turn on "pve-main" all the sensors remain at "unavailable"
When I use "reload" on the extension itself the sensors came back online (which works but has to be done manually)

For the offline node:
After I turned on the node which was unavailable during the install wizards nothing happened.
After I "reloaded" the extension all the services where discovered and added automatically.

I hope that was the test you wanted to try and that I didn't mix up something. If you need any further help just ping me pls.

**Comment by dougiteixeira** [comment]
@Munsio, Thanks for the feedback, see if you can clarify this for me:

> For the test:
> The node I am authenticated to is: "pve-asus1"
> I turned off the node "pve-main" with the extension (shutdown)
> All sensors of the VMs + Node of "pve-main" switched to "unavailable"

Do any of these "pve" node VMs have replicas set up?

The expected behavior is that if there is a replica on another node, the data should be updated by that node.

In your opinion, is the integration behavior as expected?



> When I turn on "pve-main" all the sensors remain at "unavailable"
> When I use "reload" on the extension itself the sensors came back online (which works but has to be done manually)

Did you wait a while to see if the integration would update the sensors again without reloading?

**Comment by Munsio** [comment]
> Do any of these "pve" node VMs have replicas set up?

Ah you mean the Proxmox HA setup where it moves the LXC/VM to another node. Sure I can try that one too.

> Did you wait a while to see if the integration would update the sensors again without reloading?

Will try that again. I waited for probably 2-3mins but nothing happened, will wait a little bit longer the next time.

**Comment by Munsio** [comment]
Migration of LXC did work. The LXC "node" sensor switched and all the other data like memory/cpu did also update accordingly

When waiting a few more minutes the sensors of the turned on node also switched back to their values so I do see no problems there.

> > One feature request maybe.
> > It would be great if I could set multiple hostnames to avoid a loss of connection when one of the nodes is not available. The credentials are the same for every host as I am using the "pve" realm for authentication on the API.
> 
> This is a little difficult to deal with, I'll see what's feasible.

About this one. I don't know if the python async client supports "round-robin" technique but this would definitely solve the issue where no data can be fetched if the particular node was turned off in a cluster configuration (due to update/reboot cycle maybe).
I know that this could also be solved with a reverse proxy in front but not everyone has the knowledge to do that.

**Comment by dougiteixeira** [comment, marked as answer]
> Migration of LXC did work. The LXC "node" sensor switched and all the other data like memory/cpu did also update accordingly
> 
> When waiting a few more minutes the sensors of the turned on node also switched back to their values so I do see no problems there.

Thanks for the feedback.

Since everything seems to work fine, I'll release the new version with the changes.



> About this one. I don't know if the python async client supports "round-robin" technique but this would definitely solve the issue where no data can be fetched if the particular node was turned off in a cluster configuration (due to update/reboot cycle maybe).
> I know that this could also be solved with a reverse proxy in front but not everyone has the knowledge to do that.

I'll do some testing with this and if it evolves I'll include it in a future version.
